crates/e2e/src/setup/proxy.rs

```diff
@@ -0,0 +1,133 @@
+//! Simple HTTP reverse proxy with automatic failover for e2e testing.
+//!
+//! This module provides a test-only reverse proxy that simulates how Kubernetes
+//! service pools work in production. In production, when multiple instances of
+//! a service run (e.g., autopilot with leader/follower pattern), Kubernetes
+//! routes traffic to the active instance and automatically fails over to a
+//! backup when the primary becomes unavailable.
+//!
+//! The proxy maintains two backend URLs (primary and secondary) and
+//! automatically switches between them when the currently active backend fails.
+//! This allows e2e tests to simulate production failover behavior without
+//! requiring a full k8s cluster.
+
+use {
+    axum::{Router, body::Body, http::Request, response::IntoResponse},
+    std::{net::SocketAddr, sync::Arc},
+    tokio::{sync::RwLock, task::JoinHandle},
+    url::Url,
+};
+
+/// HTTP reverse proxy with automatic failover that permanently switches
+/// to the fallback backend when the current backend fails.
+///
+/// This simulates k8s service pools where traffic is automatically routed
+/// to healthy backend instances.
+pub struct ReverseProxy {
+    _server_handle: JoinHandle<()>,
+}
+
+#[derive(Clone)]
+struct ProxyState {
+    primary: Url,
+    secondary: Url,
+    active: Arc<RwLock<ActiveBackend>>,
```

Comment:
Going the enum approach seems to introduce unnecessary matching and variables. Why not have a setup like this:
```
state ProxyState {
    backends: Arc<RwLock<VecDeque<Url>>>,
}
```
That way:
* you support arbitrarily many URLs (not needed here)
* you don't have to convert from the `ActiveBackend` to an actual URL (the active one is always the first item)
* falling back to the next URL is just popping the front and pushing it to the back

If you want to detect the case where all backends are down (e.g. for extra error handling) you just count how many backends you tried and abort when `tries = backends.len()`.